ccl/sqlproxyccl: frontend admitter

darinpp opened the pull request:
This replaces `IncomingTLSConfig` with a function similar to
`BackendDialer` that will be called to setup the frontend
connection/pipeline of the proxy.
It requires that all clients replace the calls like this one
```
  server := sqlproxyccl.NewServer(
    sqlproxyccl.Options{
      IncomingTLSConfig: &tlsConfig
    }
  }
```
with the analogous code
```
  server := sqlproxyccl.NewServer(
    sqlproxyccl.Options{
      FrontendAdmitter: func(
        incoming net.Conn,
      ) (net.Conn, *pgproto3.StartupMessage, error) {
        return sqlproxyccl.FrontendAdmit(
          incoming,
          &tls.Config,
        )
      }
    }
  }

```
While a bit more verbose, this makes possible for the proxy
library users to implement a custom logic to admit or
reject connections based on the client's IP address.

Release note: none

cc: @cockroachdb/sqlproxy-prs

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57849)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-:-MOXDQDU8bV2JaDpS9Q3:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @darinpp, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/frontend_admitter.go, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-MOX76i20t9ed3pa9lX8:-MOX76i3-DXDoN3gNNT8:b-kcch9u) ([raw file](https://github.com/cockroachdb/cockroach/blob/be50759712fe82f9b7d336c0d79133434e0321e4/pkg/ccl/sqlproxyccl/frontend_admitter.go#L18)):*
> ```Go
> )
> 
> // FrontendAdmit is an example frontend admitter
> ```

I don't think this is a good description - it's more than "an example" - it's the default admitter. Beef up this comment with information about what this method does and when and why it would be used. I'd put a bit of example code on how to use it as a library caller.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-MOXB2PeCEk8WhbPyFhM:-MOXB2PeCEk8WhbPyFhN:bpxqp17) ([raw file](https://github.com/cockroachdb/cockroach/blob/be50759712fe82f9b7d336c0d79133434e0321e4/pkg/ccl/sqlproxyccl/proxy.go#L84)):*
> ```Go
> 	// is accepted. It can optionally negotiate SSL, provide admittance control or
> 	// other types of frontend connection filtering.
> 	FrontendAdmitter func(incoming net.Conn) (net.Conn, *pgproto3.StartupMessage, error)
> ```

NIT: move this above BackendDialer since it's called first, and so I'd want to read about it first.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
Would also like to see `frontend_admitter_test.go` unit tests (missing `backend_dialer_test.go` as well, it looks like).

spaskob commented:
I think we are trying to reinvent the wheel. ie https://godoc.org/github.com/jackc/pgconn#ConnectConfig. We should use this API instead - it handles auth correctly (currently sqlproxy does not) and is going to support future changes to the protocol.

darinpp reviewed (commented):
I added several unit tests for the frontend admitter. I'll add tests for the backend dialer in another PR as I can combine with swapping it to use PG client that Spas found. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-:-MOXtDqSB_eNa8Rzvx1w:b-c4hqm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andy-kimball, @imjching, and @spaskob)

___
*[pkg/ccl/sqlproxyccl/frontend_admitter.go, line 18 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-MOX76i20t9ed3pa9lX8:-MOXt41HCWjWsxGWitmQ:b33zx3d) ([raw file](https://github.com/cockroachdb/cockroach/blob/be50759712fe82f9b7d336c0d79133434e0321e4/pkg/ccl/sqlproxyccl/frontend_admitter.go#L18)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I don't think this is a good description - it's more than "an example" - it's the default admitter. Beef up this comment with information about what this method does and when and why it would be used. I'd put a bit of example code on how to use it as a library caller.
</blockquote></details>

I expanded the comment.
___
*[pkg/ccl/sqlproxyccl/proxy.go, line 84 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57849#-MOXB2PeCEk8WhbPyFhM:-MOXt7lF4ABQgLO4IBuz:b-7mmp82) ([raw file](https://github.com/cockroachdb/cockroach/blob/be50759712fe82f9b7d336c0d79133434e0321e4/pkg/ccl/sqlproxyccl/proxy.go#L84)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: move this above BackendDialer since it's called first, and so I'd want to read about it first.
</blockquote></details>

moved


<!-- Sent from Reviewable.io -->

darinpp commented:
> I think we are trying to reinvent the wheel. ie https://godoc.org/github.com/jackc/pgconn#ConnectConfig. We should use this API instead - it handles auth correctly (currently sqlproxy does not) and is going to support future changes to the protocol.

This implements a client - I'll try replacing the Backend Dialer code to use it. This PR is more like a server side PG code.

darinpp commented:
bors r+

darinpp commented:
bors r+

darinpp commented:
bors r-

darinpp committed:
ccl/sqlproxyccl: frontend admitter

This replaces `IncomingTLSConfig` with a function similar to
`BackendDialer` that will be called to setup the frontend
connection/pipeline of the proxy.
It requires that all clients replace the calls like this one
```
  server := sqlproxyccl.NewServer(
    sqlproxyccl.Options{
      IncomingTLSConfig: &tlsConfig
    }
  }
```
with the analogous code
```
  server := sqlproxyccl.NewServer(
    sqlproxyccl.Options{
      FrontendAdmitter: func(
        incoming net.Conn,
      ) (net.Conn, *pgproto3.StartupMessage, error) {
        return sqlproxyccl.FrontendAdmit(
          incoming,
          &tls.Config,
        )
      }
    }
  }

```
While a bit more verbose, this makes possible for the proxy
library users to implement a custom logic to admit or
reject connections based on the client's IP address.

Release note: none

darinpp commented:
bors r+
